File: csharpLisp/VoidLisp.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using UnityEngine;

namespace UI {
    public class VoidLisp {
        public void TexstExe() {
            var code = @"
(list 
    (print 'hello world')
    (print (+ 1 2 ) )
    (print
         (+ 1 2 (* 3 3))
    )
)
";
            var tokens = Token(code);
            var met    = getAst(tokens).Exec();
        }


        private int          P;
        private List<string> List;

        private Met Ast() {
            var met = new Met(this);
            met.Type = Met.MetType.List;
            var tree = new List<Met>();
            met.Mets = tree;
            while (true) {
                if (P >= List.Count) {
                    Debug.LogError("tree err");
                    return null;
                }

                var m = List[P];
                P++;
                if (m == "(") {
                    var newMet = Ast();
                    tree.Add(newMet);
                }

                if (m == ")") {
                    return met;
                }

                if (m != "(" && m != ")") {
                    var notListMet = new Met(this);
                    tree.Add(notListMet);
                    var mList = m.ToCharArray();
                    if (mList[0] == '\'' && mList[mList.Length - 1] == '\'') {
                        notListMet.Type = Met.MetType.StringMet;
                        var s = "";
                        for (var i = 0; i < mList.Length; i++) {
                            if (i != 0 && i != mList.Length - 1) {
                                s = s + mList[i];
                            }
                        }

                        notListMet.MetContent = s;
                        continue;
                    }

                    notListMet.Type       = Met.MetType.BaseMat;
                    notListMet.MetContent = m;
                }
            }
        }

        //语法分析 生成语法树
        public Met getAst(List<string> li) {
            P    = 1;
            List = li;
            if (List[0] != "(") {
                Debug.LogError("tree err");
                return null;
            }

            return Ast();
        }

        //词法分析
        private List<string> Token(string exp) {
            var l1 = exp.ToList();
            var l2 = new List<string>();

            var  token      = "";
            bool isInString = false;
            for (var i = 0; i < l1.Count; i++) {
                var tc = l1[i];

                if (tc == '\'') {
                    if (!isInString && !token.Equals("")) {
                        l2.Add(token);
                    }

                    if (isInString) {
                        token = token + tc;
                        l2.Add(token);
                        token      = "";
                        isInString = false;
                        continue;
                    } else {
                        isInString = true;
                    }
                }

                if (!isInString) {
                    if (tc == ' ' || tc == '\r' || tc == '\n') {
                        if (!token.Equals("")) {
                            l2.Add(token);
                        }

                        token = "";
                        continue;
                    }

                    if (tc == '(' || tc == ')') {
                        if (!token.Equals("")) {
                            l2.Add(token);
                        }

                        l2.Add(tc.ToString());
                        token = "";
                        continue;
                    }
                }

                token = token + tc.ToString();
            }

            return l2;
        }

        public Met Eval(string code) {
            var tokens = Token(code);
            var met    = getAst(tokens).Exec();
            return met;
        }


        public Dictionary<string, LispFunction> Functions;
        public Dictionary<string, Met> Values =  new Dictionary<string, Met>();

        public VoidLisp() {
            Functions = new Dictionary<string, LispFunction> {
                {
                    "list",
                    new LispFunction(li => {
                        var listMet = new Met(this);
                        listMet.Type       = Met.MetType.List;
                        listMet.Mets       = li;
                        listMet.MetContent = "list";
                        return listMet;
                    })
                },
                {"eval", new LispFunction(li => Eval(li[0].MetContent))},
                {
                    "let" , 
                    new LispFunction(li => {
                        Values.Add(li[0].MetContent,li[1]);
                        var met = new Met(this);
                        met.Type = Met.MetType.BaseMat;
                        met.MetContent = "T";
                        return met;
                    })
                },
                {
                    "print", new LispFunction(l => {
                        var s = "";
                        foreach (var met in l) {
                            met.Print();
                            s = s + " " + met.MetContent;
                        }

                        var resmet = new Met(this);
                        resmet.MetContent = s;
                        resmet.Type       = Met.MetType.BaseMat;
                        return resmet;
                    })
                },
                {
                    "quote", new LispFunction(metList => {
                        var met = new Met(this);
                        met.Type = Met.MetType.List;
                        met.Mets = metList;
                        return met;
                    })
                },
                {
                    "+", new LispFunction(metList => {
                        var sum = 0f;
                        foreach (var met in metList) {
                            var n = float.Parse(met.MetContent);
                            sum = n + sum;
                        }

                        var m = new Met(this);
                        m.Type       = Met.MetType.BaseMat;
                        m.MetContent = sum.ToString();
                        return m;
                    })
                },
                {
                    "*", new LispFunction(metList => {
                        var sum = 1f;
                        foreach (var met in metList) {
                            var n = float.Parse(met.MetContent);
                            sum = n * sum;
                        }

                        var m = new Met(this);
                        m.Type       = Met.MetType.BaseMat;
                        m.MetContent = sum.ToString();
                        return m;
                    })
                },
                {
                    "-", new LispFunction(metList => {
                        var result = float.Parse(metList[0].MetContent);
                        for (var i = 1; i < metList.Count; i++) {
                            var n = float.Parse(metList[i].MetContent);
                            result = result - n;
                        }

                        var m = new Met(this);
                        m.Type       = Met.MetType.BaseMat;
                        m.MetContent = result.ToString();
                        return m;
                    })
                },
                {
                    "/", new LispFunction(metList => {
                        var result = float.Parse(metList[0].MetContent);
                        for (var i = 1; i < metList.Count; i++) {
                            var n = float.Parse(metList[i].MetContent);
                            result = result / n;
                        }

                        var m = new Met(this);
                        m.Type       = Met.MetType.BaseMat;
                        m.MetContent = result.ToString();
                        return m;
                    })
                },
                {
                    "loop", new LispFunction(li => {
                        return null;
                    })
                },
                {
                    "openMsgW" , new LispFunction(l => {
                        UIUtils.SimpleMsgWindow(l[0].MetContent, () => {
                        }, () => {
                        });
                        var m = new Met(this);
                        m.Type = Met.MetType.BaseMat;
                        
                        return m;
                    })
                }
                
            };
        }
    }


    public class Met {
        private VoidLisp _eviroment;

        public Met(VoidLisp eviroment) {
            _eviroment = eviroment;
        }

        public enum MetType {
            BaseMat,
            List,
            StringMet,
            Number
        }

        public string    Token;
        public List<Met> Mets;
        public MetType   Type;
        public string    MetContent;

        public Met Exec() {
            switch (Type) {
                case MetType.BaseMat:
                    if (_eviroment.Values.ContainsKey(MetContent)) {
                        return _eviroment.Values[MetContent];
                    } else {
                        return this;
                    }
                case MetType.List:
                    var n1        = Mets[0].Exec();
                    var paramList = new List<Met>();
                    for (var i = 0; i < Mets.Count; i++) {
                        if (i == 0) {
                            continue;
                        }

                        var li    = Mets[i];
                        var param = li.Exec();
                        paramList.Add(param);
                    }

                    var result = _eviroment.Functions[n1.MetContent].Function.Invoke(paramList);
                    return result;
                case MetType.StringMet:
                    return this;
                case MetType.Number:
                    return this;
                default:
                    return this;
            }
        }

        public void Print() {
            switch (Type) {
                case MetType.BaseMat:
                    Debug.Log(MetContent);
                    break;
                case MetType.List:
                    Debug.Log("---------------");
                    Mets.ForEach(m => m.Print());
                    break;
                case MetType.StringMet:
                    Debug.Log(MetContent);
                    break;
                case MetType.Number:
                    break;
                default:
                    throw new ArgumentOutOfRangeException();
            }
        }
    }

    public class LispFunction {
        public Func<List<Met>, Met> Function;

        public LispFunction(Func<List<Met>, Met> function) {
            Function = function;
        }
    }
}

File: jsLisp/VoidLisp.js
"use strict";
exports.__esModule = true;
var VoidLisp = /** @class */ (function () {
    function VoidLisp() {
        var _this = this;
        this.Functions = {
            "q": function (li) {
                return li[0];
            },
            "exec": function (li) {
                return li[0].Exec();
            },
            "list": function (li) {
                var met = new Met(_this, MetType.List);
                met.Mets = li;
                met.MetContent = "list";
                return met;
            },
            "push": function (li) {
                var met = new Met(_this, MetType.List, "list");
                li[1].Mets.push(li[0]);
                met.Mets = li[1].Mets;
                console.log(met);
                return met;
            },
            "index": function (li) {
                var list = li[1];
                var index = li[0];
                var met = li[1].Mets[parseInt(li[0].MetContent)];
                if (met === undefined) {
                    throw "list out of range ,inde:" + parseInt(li[0].MetContent);
                }
                return met;
            },
            "print": function (li) {
                var s = "";
                for (var i = 0; i < li.length; i++) {
                    s = s + li[i].MetContent;
                }
                var met = new Met(_this, MetType.BaseMat, s);
                console.log(s);
            },
            "or": function (li) {
                var met = new Met(_this, MetType.BaseMat, "F");
                for (var i = 0; i < li.length; i++) {
                    if (li[i].MetContent === "T") {
                        return new Met(_this, MetType.BaseMat, "T");
                    }
                }
                return met;
            },
            "and": function (li) {
                var met = new Met(_this, MetType.BaseMat, "T");
                for (var i = 0; i < li.length; i++) {
                    if (li[i].MetContent === "F") {
                        return new Met(_this, MetType.BaseMat, "F");
                    }
                }
                return met;
            },
            "not": function (li) {
                return li[0].MetContent === "F" ? new Met(_this, MetType.BaseMat, "T") : new Met(_this, MetType.BaseMat, "F");
            },
            "==": function (li) {
                if (li[0].MetContent === li[1].MetContent) {
                    return new Met(_this, MetType.BaseMat, "T");
                }
                else {
                    return new Met(_this, MetType.BaseMat, "F");
                }
            },
            "<": function (li) {
                var f1 = parseFloat(li[0].MetContent);
                var f2 = parseFloat(li[1].MetContent);
                if (f1 < f2) {
                    return new Met(_this, MetType.BaseMat, "T");
                }
                else {
                    return new Met(_this, MetType.BaseMat, "F");
                }
            },
            ">": function (li) {
                var f1 = parseFloat(li[0].MetContent);
                var f2 = parseFloat(li[1].MetContent);
                if (f1 > f2) {
                    return new Met(_this, MetType.BaseMat, "T");
                }
                else {
                    return new Met(_this, MetType.BaseMat, "F");
                }
            },
            "+": function (li) {
                var first = parseFloat(li[0].MetContent);
                for (var i = 1; i < li.length; i++) {
                    var n = parseFloat(li[i].MetContent);
                    first = first + n;
                }
                var met = new Met(_this, MetType.BaseMat, first + "");
                return met;
            },
            "-": function (li) {
                var first = parseFloat(li[0].MetContent);
                for (var i = 1; i < li.length; i++) {
                    var n = parseFloat(li[i].MetContent);
                    first = first - n;
                }
                var met = new Met(_this, MetType.BaseMat, first + "");
                return met;
            },
            "*": function (li) {
                var first = parseFloat(li[0].MetContent);
                for (var i = 1; i < li.length; i++) {
                    var n = parseFloat(li[i].MetContent);
                    first = first * n;
                }
                var met = new Met(_this, MetType.BaseMat, first + "");
                return met;
            },
            "/": function (li) {
                var first = parseFloat(li[0].MetContent);
                for (var i = 1; i < li.length; i++) {
                    var n = parseFloat(li[i].MetContent);
                    first = first / n;
                }
                var met = new Met(_this, MetType.BaseMat, first + "");
                return met;
            },
            "if": function (li) {
                var isT = li[0];
                if (isT.MetContent === "T") {
                    return li[1].Exec();
                }
                else {
                    return li[2].Exec();
                }
            },
            "while": function (li) {
                while (li[0].MetContent === "T") {
                    var p = li[1].Exec();
                }
                return new Met(_this);
            },
            "def": function (li) {
                //value
                var v = li[1];
                _this.Functions[li[0].MetContent] = function (lii) { return v; };
                return new Met(_this);
            },
            "lambda": function (li) {
                if (li[1].Type != MetType.List) {
                    throw "no paramLis:t " + li[0].MetContent;
                }
                //Function
                var paramList = li[1].Mets;
                _this.Functions[li[0].MetContent] = function (lii) {
                    //give memery before function invoke
                    for (var i = 0; i < paramList.length; i++) {
                        var p = lii[i];
                        _this.Functions[paramList[i].MetContent] = function (l) { return p; };
                    }
                    //TODO: release memery when function invoked
                    return li[2].Exec();
                };
                return new Met(_this);
            }
        };
    }
    VoidLisp.prototype.TestExe = function () {
        var code = "\n\t\t(list \n\t\t\t(print 'hello world')\n\t\t\t(print (+ 1 2 ) )\n\t\t\t(print\n\t\t\t\t(+ 1 2 (* 3 3))\n\t\t\t)\n\t\t)\n\t\t ";
        var tokens = this.Token(code);
        var met = this.getAst(tokens).Exec();
    };
    VoidLisp.prototype.Ast = function () {
        var met = new Met(this, MetType.List, "list");
        met.Type = MetType.List;
        var tree = new Array();
        met.Mets = tree;
        while (true) {
            if (this.P >= this.List.length) {
                console.log("tree err");
                return null;
            }
            var m = this.List[this.P];
            this.P = this.P + 1;
            if (m === "(") {
                var newMet = this.Ast();
                tree.push(newMet);
            }
            if (m === ")") {
                return met;
            }
            if (m != "(" && m != ")") {
                var notListMet = new Met(this);
                tree.push(notListMet);
                var mList = m;
                if (mList[0] == '"' && mList[mList.length - 1] == '"') {
                    notListMet.Type = MetType.StringMet;
                    var s = "";
                    for (var i = 0; i < mList.length; i++) {
                        if (i != 0 && i != mList.length - 1) {
                            s = s + mList[i];
                        }
                    }
                    notListMet.MetContent = s;
                    continue;
                }
                notListMet.Type = MetType.BaseMat;
                notListMet.MetContent = m;
            }
        }
    };
    //语法分析 生成语法树
    VoidLisp.prototype.getAst = function (li) {
        this.P = 1;
        this.List = li;
        if (this.List[0] != "(") {
            console.log("tree err");
            return null;
        }
        return this.Ast();
    };
    //词法分析
    VoidLisp.prototype.Token = function (exp) {
        var l1 = exp;
        var l2 = new Array();
        var token = "";
        var isInString = false;
        var isInAno = false;
        for (var i = 0; i < l1.length; i++) {
            var tc = l1[i];
            if (tc === ';') {
                isInAno = true;
            }
            if (tc === '\r' || tc === '\n') {
                isInAno = false;
            }
            if (isInAno) {
                continue;
            }
            if (tc === '"') {
                if (!isInString && token !== "") {
                    l2.push(token);
                }
                if (isInString) {
                    token = token + tc;
                    l2.push(token);
                    token = "";
                    isInString = false;
                    continue;
                }
                else {
                    isInString = true;
                }
            }
            if (!isInString) {
                if (tc === ' ' || tc === '\r' || tc === '\n' || tc === "\t") {
                    if (token !== "") {
                        l2.push(token);
                    }
                    token = "";
                    continue;
                }
                if (tc === '(' || tc === ')') {
                    if (token !== "") {
                        l2.push(token);
                    }
                    l2.push(tc);
                    token = "";
                    continue;
                }
            }
            token = token + tc;
        }
        return l2;
    };
    VoidLisp.prototype.Eval = function (code) {
        var tokens = this.Token(code);
        var met = this.getAst(tokens).Exec();
        return met;
    };
    return VoidLisp;
}());
exports.VoidLisp = VoidLisp;
var MetType;
(function (MetType) {
    MetType[MetType["BaseMat"] = 0] = "BaseMat";
    MetType[MetType["List"] = 1] = "List";
    MetType[MetType["StringMet"] = 2] = "StringMet";
    MetType[MetType["NumberMet"] = 3] = "NumberMet";
})(MetType || (MetType = {}));
var Met = /** @class */ (function () {
    function Met(eviroments, types, content) {
        if (types === void 0) { types = MetType.BaseMat; }
        if (content === void 0) { content = "Nil"; }
        this.Type = types;
        this.eviroment = eviroments;
        this.MetContent = content;
    }
    Met.prototype.Exec = function () {
        switch (this.Type) {
            case MetType.BaseMat:
                if (this.eviroment.Functions[this.MetContent] != undefined) {
                    return this.eviroment.Functions[this.MetContent]().Exec();
                }
                else {
                    return this;
                }
            case MetType.List:
                var n1 = this.Mets[0];
                var paramList = new Array();
                for (var i = 0; i < this.Mets.length; i++) {
                    if (i == 0) {
                        continue;
                    }
                    var li = this.Mets[i].Exec();
                    var param = li;
                    paramList.push(param);
                }
                var result = this.eviroment.Functions[n1.MetContent](paramList);
                return result;
            case MetType.StringMet:
                return this;
            case MetType.NumberMet:
                return this;
        }
    };
    Met.prototype.Print = function () {
        console.log(this);
        switch (this.Type) {
            case MetType.BaseMat:
                console.log(this.MetContent);
                break;
            case MetType.List:
                console.log("---------------");
                break;
            case MetType.StringMet:
                console.log(this.MetContent);
                break;
            case MetType.NumberMet:
                break;
        }
    };
    return Met;
}());
var code = "\n(runlist (print 'hello world')\n\n\t(print ( or  (< 1 -2) (> 2 3)))\n\t(print ( < -12 (+ 0 1)))\n\n\t(let var 15)\n\t(print var)\n\t(let var 'hello world')\n\t(print var)\n\n\t(def log (s) (print s))\n\t(log 'test log')\n\n\t(let i 1)\n\t(while (< i 100) \n\t\t(runlist (print i)\n\t\t\t(let i (+ i 1))))\n\n\t(print 'end'))\n\n";
